### PR TITLE
New docs, refactoring

### LINKED ISSUE #50: API documentation
As the project begins to stabilize, it is worth reviewing and documenting (possibly also refactoring) the modules used by bot developers.

Some proper guides would also be helpful:
- "Step by step: write a simple bot"
- "Custom battles: how to run battles without the GUI"

### PR DESCRIPTION
**Pull request by ArielHorwitz**
Adds a new run option to automatically create html docs (with guides) locally from source. Many new docstrings and type hinting. Some API changes - this should mark a more conservative/stable API for the future.

Addresses #50.

### CODE REVIEW

File: bots/__init__.py [hunk 1]
```diff
@@ -45,7 +63,7 @@ def get_bot_classes(
     # Count slots
     total_bots = len(available_bots)
     if total_bots == 0:
-        raise NotFairError(f'Found 0 bots to fill {total_slots} slots. Selection: {selection} - {ignore} (minus testing: {not include_testing})')
+        raise NotFairError(f'Found <2 bots to fill {total_slots} slots. Selection: {selection} - {ignore} (minus testing: {not include_testing})')
```
Comment (imcrazeegamer) [inline bots/__init__.py:66]: "that string is incorrect,
damn it ill fix that" - python developer

File: run/makedocs.py [hunk 2]
```diff
@@ -0,0 +1,176 @@
...
+        main_doc_strs.extend(categories[c])
+    # Finalize and write
+    main_doc_strs.append('"""')
+    main_doc = '\n'.join(main_doc_strs)
+    file_dump(GUIDES_PY_DIR / '__init__.py', main_doc, clear=True)
+
+
+def _windows_compat(s):
+    """Add backslashes for windows path compatibility for `file_dump`."""
+    return s.replace('\\', '\\\\')
```
Comment (imcrazeegamer) [inline run/makedocs.py:131]:
```
                                                                                          
88                                88                  88                      88           
88                                88                  88                      88           
88                                88                  88                      88           
88,dPPYba,  ,adPPYYba,  ,adPPYba, 88   ,d8  ,adPPYba, 88 ,adPPYYba, ,adPPYba, 88,dPPYba,   
88P'    "8a ""     `Y8 a8"     "" 88 ,a8"   I8[    "" 88 ""     `Y8 I8[    "" 88P'    "8a  
88       d8 ,adPPPPP88 8b         8888[      `"Y8ba,  88 ,adPPPPP88  `"Y8ba,  88       88  
88b,   ,a8" 88,    ,88 "8a,   ,aa 88`"Yba,  aa    ]8I 88 88,    ,88 aa    ]8I 88       88  
8Y"Ybbd8"'  `"8bbdP"Y8  `"Ybbd8"' 88   `Y8a `"YbbdP"' 88 `"8bbdP"Y8 `"YbbdP"' 88       88  
                                                                                           
                                                                                           
```

File: run/makedocs.py [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+def _find_guide_title_categories(guide_md_file):
+    markdown = file_load(guide_md_file)
+    title = None
+    categories = []
+    for line in markdown.split('\n'):
+        if not title and line.startswith('# '):
+            title = line[2:]
+        if not categories:
+            if line.startswith('*Categories: ') and line.endswith('*'):
+                categories = line[13:-1].split(', ')
```
Comment (imcrazeegamer) [inline run/makedocs.py:143]: Caution with magic numbers

File: util/file.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+Writing, loading, and opening files.
+"""
+import os
+import subprocess
+import platform
+
+
+def file_load(file: os.PathLike):
+    """Loads *file* and returns the contents as a string."""
+    with open(file, 'r') as f:
```
Comment (imcrazeegamer) [inline util/file.py:11]: might want to set the same encoding on load and save

**Review by imcrazeegamer — APPROVED** [review]
The Opposition is Victorious,
Long Live The Doc Strings!!

File: bots/__init__.py (hunk 1 above)
Comment (ArielHorwitz) [reply bots/__init__.py:66]: On fixlist

File: run/makedocs.py (hunk 3 above)
Comment (ArielHorwitz) [reply run/makedocs.py:143]: on fixlist

File: util/file.py (hunk 4 above)
Comment (ArielHorwitz) [reply util/file.py:11]: Duely noted. Will do testing. Prefer to see an error before fix.

File: run/makedocs.py (hunk 2 above)
Comment (ArielHorwitz) [reply run/makedocs.py:131]: \\\\\\\\\\\\\\\\\\

Is that enough??

**Pull request merged**